### PR TITLE
Introduce publishing to DWP Find a Job service

### PR DESCRIPTION
**Pull request by scruti**
## Trello card URL
https://trello.com/c/MGIXSJIZ

## Changes in this PR:
Introduce services that allow us to publish/update and expire our internal (not imported) published vacancies to the DWP Find a Job service.

The integration with "Find a Job" service is done through an SFTP connection, publishing XML files that they will import daily into their service.

We will publish:
- one file for publishing/editing and 
- one file for expiring/deleting the vacancies.

**Documentation:**
You can see the documentation for the integration on: https://github.com/DFE-Digital/teaching-vacancies/blob/c7f9e696aef6fe92d1242de421b4b560a61dff50/documentation/integrations/dwp-find-a-job.md

image

## Missing bits/questions:
- **This is not scheduling the Jobs to run periodically**
That will be done in a new PR once we manually execute the publishing to test its behaviour.
DWP Find a Job doesn't have a test environment. So we need to do test runs directly against production. Fun, eh? :D

- **We are not sending the Salary**
  Salary is optional in the DWP Find a Job service. Since it requires exact values for min/max salary and a period, but we store it as a string with numbers and words, we haven't included salary parsing. I'm open to implementing this but it will be flaky and prone to edge cases causing wrong values in DWP.

- **What happens when we hard-delete a vacancy in our system?** 
  We send the vacancies marked as "early-closed" by hiring staff to be deleted. But if we hard-delete a vacancy from DB, that means the vacancy (If already published) will stay in DWP until its expiration date arrives, while pointing to a dead link in our system.
***How could we solve this?***
We have two main caveats:
  - We need to send the `id` of the deleted vacancies in an XML. And that becomes unavailable with hard-deleting.
  - If deleting multiple vacancies, we do not want to generate/push an XML for each one (even if that would work) but generate an XML for all of them.

  A possible solution would be to add some trigger/callback when hard deleting the vacancy that will publish an "event" adding the vacancy `id` to Redis memory. And that would be regularly read by a scheduled job that would generate a deletion file and push it to DWP when finding "ids to delete" in the service.
  We could also implement this with a new DB table instead of memory. Need to figure out the pros/cons.

### CODE REVIEW

File: app/services/vacancies/export/dwp_find_a_job/expired_and_deleted/query.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
+module Vacancies::Export::DwpFindAJob::ExpiredAndDeleted
+  class Query
+    attr_reader :from_date
+
+    def initialize(from_date)
+      @from_date = from_date
+    end
+
+    def vacancies
+      @vacancies ||= vacancies_updated_to_expire_on_date
```
[the remark this answers is not in the document]
Comment (scruti) [reply app/services/vacancies/export/dwp_find_a_job/expired_and_deleted/query.rb:10]: Good point! Will re-visit this (I don't even think we need to cache the query).

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -35,8 +35,9 @@
         "sleistner.vscode-fileutils",
         "Shopify.ruby-extensions-pack",
         "GitHub.copilot",
-        "donjayamanne.git-extension-pack"
-	],
+        "donjayamanne.git-extension-pack",
+        "bierner.markdown-mermaid"
```
Comment (scruti) [inline .devcontainer/devcontainer.json:39]: This allows to preview mermaid markdown graphs within the devcontainer.

**Review by KyleMacPherson — APPROVED** [review]

File: documentation/integrations/dwp-find-a-job.md [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+```
+
+## Uploading new/edited vacancies
+
+The service orcherstating the upload of new/edited vacancies is [Vacancies::Export::DwpFindAJob::NewAndEdited::Upload](../../app/services/vacancies/export/dwp_find_a_job/new_and_edited/upload.rb)
+
+### What vacancies are exported?
+We publish vacancies that fall on any of these conditions:
+- Got published on/after the given date.
+  EG: Yesterday a date, all the vacancies published yesterday and today will be returned.
```
[the remark this answers is not in the document]
Comment (scruti) [reply documentation/integrations/dwp-find-a-job.md]: Good catch! Too much text editing! Fixed!

**Comment by scruti** [comment]
Merging the PR so we can do manual runs on the server for testing the integration end-to-end.
Remember DWP Find a Job doesn't have testing environment, so we need to test this against PROD :see_no_evil:

**Pull request merged**